packages/salesforcedx-vscode-core/src/commands/forceSourceDeploySourcePath.ts

```diff
@@ -130,100 +133,102 @@ export async function forceSourceDeployMultipleSourcePaths(uris: vscode.Uri[]) {
   await commandlet.run();
 }
 
-export class LibraryDeploySourcePathExecutor extends DeployRetrieveLibraryExecutor {
-  private hashElement(component: SourceComponent): string {
-    const hashed = `${component.fullName}.${component.type.id}`;
-    return hashed;
-  }
+export class LibraryDeploySourcePathExecutor extends LibraryCommandletExecutor<
+  string | string[]
+> {
+  protected executionName = 'Deploy (Beta)';
+  protected logName = 'force_source_deploy_with_sourcepath_beta';
+  private diagnostics = vscode.languages.createDiagnosticCollection(
+    'deploy-errors'
+  );
 
-  public async execute(
+  public async run(
     response: ContinueResponse<string | string[]>
-  ): Promise<void> {
-    this.setStartTime();
+  ): Promise<boolean> {
     try {
-      await this.build(
-        'Deploy (Beta)',
-        'force_source_deploy_with_sourcepath_beta'
-      );
-      channelService.showCommandWithTimestamp(`Starting ${this.executionName}`);
-
-      let components: SourceComponent[] = [];
-      const registryAccess = new RegistryAccess();
-      if (typeof response.data === 'string') {
-        components = registryAccess.getComponentsFromPath(response.data);
-      } else {
-        const allComponents: SourceComponent[] = [];
-
-        for (const filepath of response.data) {
-          allComponents.push(...registryAccess.getComponentsFromPath(filepath));
-        }
-
-        const hashedCmps = new Set();
-        components = allComponents.filter(component => {
-          const hashed = this.hashElement(component);
-          if (!hashedCmps.has(hashed)) {
-            hashedCmps.add(hashed);
-            return component;
-          }
-        });
-      }
-
-      const projectNamespace = (await SfdxProjectConfig.getValue(
+      const getConnection = WorkspaceContext.get().getConnection();
+      const components = this.getComponents(response.data);
+      const namespace = (await SfdxProjectConfig.getValue(
         'namespace'
       )) as string;
 
-      if (this.sourceClient === undefined) {
-        throw new Error('SourceClient is not established');
-      }
-
-      let deploy: Promise<SourceDeployResult>;
-      let api: string;
-      if (projectNamespace) {
-        deploy = this.sourceClient.tooling.deploy(components, {
-          namespace: projectNamespace
-        });
-        api = 'tooling';
-      } else {
-        deploy = this.sourceClient.metadata.deploy(components);
-        api = 'metadata';
-      }
-
+      const deploy = this.doDeploy(await getConnection, components, namespace);
       const metadataCount = JSON.stringify(createComponentCount(components));
-      const result = await vscode.window.withProgress(
-        {
-          title: this.executionName,
-          location: vscode.ProgressLocation.Notification
-        },
-        () => deploy
-      );
+      this.telemetry.addProperty('metadataCount', metadataCount);
```

Comment:
If commands want to add extra telemetry data, call `addProperty`/`addMeasurement` using the base class' `TelemetryBuilder`